### ISSUE TITLE
Pyright marks incorrect type for sequence pattern matching

### ISSUE THREAD

**Issue by StockerMC** [microsoft/pyright#2439]
**Screenshots or Code**
```python
from typing import TYPE_CHECKING, TypeVar
from collections.abc import Callable

A = TypeVar('A', bound=int)
B = TypeVar('B')

if not TYPE_CHECKING:
    reveal_type = lambda x: None

# if we replace B with Any, result and next_arg will be Any
def get(func: Callable[[A], B], start: A) -> B:
    match func(start):
        case None:
            return get(func, start + 1)
        case (result, next_arg):
            # 600 7
            print(result, next_arg)
            reveal_type(result)
            reveal_type(next_arg)
            # Expression of type "Never" cannot be assigned to return type "B@get"
            return result
        case _:
            raise RuntimeError

# 600
print(get(lambda a: None if a <= 5 else (a * 100, a + 1), 5))

collection: list[int | list[int]] = [2, [2, 3, 4]]

match collection:
    case 1, [x, *others]:
        # Type of "others" is "tuple[int, ...]"
        # It should be "list[int]"
        reveal_type(others)
        # with collection: list[int | list[int]] = [1, [2, 3, 4]]
        # others == [3, 4]
        print(f"Got 1 and a nested sequence: {others}")
    case (a, b):
        # both a and b are "int | list[int] which is correct
        print(f"Got 2 items, {a} and {b}")
    case (1, x):
        # x is "int | list[int]" which is correct
        print(f"Got 1 and {x}")
```
```
  info: Type of "result" is "Never"
  info: Type of "next_arg" is "Never"
  error: Expression of type "Never" cannot be assigned to return type "B@get"
    Type "Never" cannot be assigned to type "B@get" (reportGeneralTypeIssues)
  info: Type of "others" is "tuple[int, ...]"
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line (pyright 1.1.178)

**Comment by erictraut** [comment]
I think this is expected behavior. The type of the expression `func(start)` is `B@get`, and this type variable is not bound to a type that derives from `Sequence`, so it will never match a sequence pattern `(result, next_arg)`.

If I change the signature of `get` to the following, it works. Is this perhaps what you intended?

```python
def get(func: Callable[[A], Sequence[B]], start: A) -> B:
```

**Comment by StockerMC** [comment]
Ah yeah, I missed that, thanks. I do believe that the second example should have the type of `others` match the collection type (in this case, `list[int]`, though.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.179, which I just published. It will also be included in the next release of pylance.
